java-api-core/src/com/kinvey/java/store/BaseDataStore.java

```diff
@@ -457,104 +460,96 @@ public KinveyPullResponse pullBlocking(Query query, boolean isAutoPagination) th
+            response.setCount(readResponse.getResult().size());
+        } else {
+            if (query.getSortString() == null || query.getSortString().isEmpty()) {
+                query.addSort(Constants._ID, AbstractQuery.SortOrder.ASC);
+            }
+            List<Exception> exceptions = new ArrayList<>();
+            int skipCount = 0;
+
+            // First, get the count of all the items to pull
+            int totalItemNumber = countNetwork();
+            String lastRequestTime = null;
+            int pulledItemCount = 0;
+            int totalPagesNumber = Math.abs(totalItemNumber / pageSize) + 1;
+            int batchSize = BATCH_SIZE; // batch size for concurrent push requests
+            ExecutorService executor;
+            List<FutureTask<PullTaskResponse>> tasks;
+            NetworkManager.Get pullRequest;
+            FutureTask<PullTaskResponse> ft;
+            for (int i = 0; i < totalPagesNumber; i += batchSize) {
+                executor = Executors.newFixedThreadPool(batchSize);
+                tasks = new ArrayList<>();
+                do {
+                    query.setSkip(skipCount).setLimit(pageSize);
+                    pullRequest = networkManager.pullBlocking(query);
+                    skipCount += pageSize;
+                    try {
+                        ft = new FutureTask<PullTaskResponse>(new CallableAsyncPullRequestHelper(pullRequest, query));
+                        tasks.add(ft);
+                        executor.execute(ft);
+                    } catch (AccessControlException | KinveyException e) {
+                        e.printStackTrace();
+                        exceptions.add(e);
+                    } catch (Exception e) {
+                        throw e;
+                    }
+                } while (skipCount < totalItemNumber);
+
+                for (FutureTask<PullTaskResponse> task : tasks) {
+                    try {
+                        PullTaskResponse tempResponse = task.get();
+                        cache.delete(tempResponse.getQuery());
+                        pulledItemCount += cache.save(tempResponse.getKinveyReadResponse().getResult()).size();
+                        exceptions.addAll(tempResponse.getKinveyReadResponse().getListOfExceptions());
+                        lastRequestTime = tempResponse.getKinveyReadResponse().getLastRequestTime();
+                    } catch (InterruptedException | ExecutionException e) {
+                        e.printStackTrace();
+                    }
                 }
+                executor.shutdown();
+            }
+            response.setListOfExceptions(exceptions);
+            response.setCount(pulledItemCount);
+            if (deltaSetCachingEnabled && lastRequestTime != null) {
+                queryCache.save(new QueryCacheItem(
+                        getCollectionName(),
+                        query.getQueryFilterMap().toString(),
+                        lastRequestTime));
             }
```

Comment:
When performing auto-pagination, is this guaranteed to be the earliest `lastRequestTime` of the numerous `GET` requests sent?

Reply:
It will be the time from the last request which was executed.

Reply:
Based on the design doc, we should set the `lastRequestTime` to the request time returned from the `_count` request made prior to the auto-paginated GET requests:

https://github.com/Kinvey/java-library/blob/072abd9876591a999f036fd66367069c6912a02c/java-api-core/src/com/kinvey/java/store/BaseDataStore.java#L487